Adding support for Ubunt 16.04 and Centos 7.

hyeomans opened the pull request:
Right now this cookbook will build `mcrouter` from source. Once this cookbook is merged it will be really easy to install `mcrouter` from packages. I didn't want to include that change in here, because there is already a lot of changes on this PR.

This cookbook is a translation of the bash script I made a couple of weeks ago but adding Chef specific constructs.

The end result is having a node with `mcrouter` running on it.

Brantone commented on CHANGELOG.md:
```diff
@@ -1,5 +1,16 @@
+### Breaking changes:
+* Removed support for Ubuntu 14.04
+* Converted `node['folly']['version']` to `node['mcrouter']['commit_hash']` given that FB has a weird release process.
+* Removed `node['folly']['version']` given that the commit has is part of mcrouter project. See [FOLLY_COMMIT](https://github.com/facebook/mcrouter/blob/release-36-0/mcrouter/FOLLY_COMMIT)
```
Converted and removed? If it's converted, doesn't that imply it no longer exists, aka: similar to removed?

hyeomans replied:
I think what I meant is "removed the ability to specify the Folly commit hash as an attribute" does that make sense?

Brantone replied:
Current wording suggest attr rename, if it's functionality conversion, should be worded as such.
That being said, see other comment about not being able to specify specific folly version.

Brantone commented on README.md:
```diff
@@ -1,17 +1,32 @@
-# mcrouter
-[![Build Status](https://travis-ci.org/evertrue/mcrouter-cookbook.svg)](https://travis-ci.org/evertrue/mcrouter-cookbook)
```
Reason for removal?
It's quite common for cookbooks to retain build status at top of Readme

hyeomans replied:
It wasn't working anymore, the original maintainer is aware of this and he's trying to fix it on his own branch.

Brantone replied:
Is it that it's broken due to code, or due to pipeline setup? 
Last run (https://travis-ci.org/evertrue/mcrouter-cookbook/jobs/265735410) had some failure logs with it.

hyeomans replied:
kitchen runs fine I think something changed on the way it connected to travis-ci. I think the original maintainer is trying to work it out on the master branch.

Brantone commented on recipes/initial_setup.rb:
```diff
@@ -1,8 +1,9 @@
 #
 # Cookbook Name:: mcrouter
-# Spec:: service
+# Recipe:: double_conversion
```
This already exists in `recipes/double_conversion.rb`, does it not?

Brantone commented on spec/unit/recipes/default_spec.rb:
```diff
@@ -19,19 +20,20 @@
 require 'spec_helper'
 
 describe 'mcrouter::default' do
+<<<<<<< HEAD
```
Merge conflict?

Brantone commented on recipes/initial_setup.rb:
```diff
@@ -1,8 +1,9 @@
 #
```
With the change in location and name, means losing out on a unit test?

Brantone commented on CHANGELOG.md:
```diff
@@ -1,5 +1,16 @@
+
+### Breaking changes:
+* Removed support for Ubuntu 14.04
+* Converted `node['folly']['version']` to `node['mcrouter']['commit_hash']` given that FB has a weird release process.
```
What's weird about the release process?
Looking at https://github.com/facebook/folly/releases it's appears to be weekly release. By removing version, and not having a hash for the `folly` repo, does that mean it's getting HEAD? Typically not advisable unless good reason.

hyeomans replied:
Oh ok, so the release process I was referring to was `mcrouter`. You can see that `mcrouter` includes a FOLLY_COMMIT and WANGLE_COMMIT file in its repo:

https://github.com/facebook/mcrouter/commit/e907287079697baed001eedcac1194d7eeb86991

So the build process, according to their own scripts, is to read that commit hash and checkout out that specific version, you can see that here for Folly
https://github.com/facebook/mcrouter/blob/master/mcrouter/scripts/recipes/folly.sh#L8
and here for Wangle:
https://github.com/facebook/mcrouter/blob/master/mcrouter/scripts/recipes/wangle.sh#L8

Removing that `[folly][version]` attribute had the idea of "as a user of this cookbook you don't have to worry about the commit hashes, because Facebook already provides that in their own mcrouter repo" if that makes sense.

hyeomans replied:
@Brantone @mitch-roblox I'm just going to put back the attributes with the commit hashes and update the README clarifiying that if you want to update mcrouter version don't forget to check if `folly`, `double-conversion` and `wangle` commit hashes have changed.

mitch-roblox replied:
Sounds good.

Brantone commented on recipes/service.rb:
```diff
@@ -16,14 +17,14 @@
-  supports status: true, restart: true
-  action [:start, :enable]
-  subscribes :restart, 'template[/etc/init/mcrouter.conf]'
+systemd_service 'mcrouter' do
```
Since adding `systemd`, probably wanna call that out in the Changelog.

hyeomans replied:
Updated it

Brantone commented on recipes/wangle.rb:
```diff
@@ -0,0 +1,48 @@
+
+git "#{Chef::Config[:file_cache_path]}/wangle" do
+  repository 'https://github.com/facebook/wangle'
+  revision lazy { File.open("#{Chef::Config[:file_cache_path]}/mcrouter/mcrouter/WANGLE_COMMIT", &:readline) }
```
This is dependent on that file even existing though, right?

hyeomans replied:
I'm going to put the commit hash as an attribute.

Brantone commented on recipes/_deps_centos.rb:
```diff
@@ -0,0 +1,112 @@
+  package pkg
+end
+
+remote_file "#{cache_dir}/libarchive-3.1.2-10.el7_2.x86_64.rpm" do
```
How about an attribute for the version?

Brantone replied:
`remote_file` should also always have a checksum to ensure integrity.

mitch-roblox replied:
Agree with both of Brenton's comments here.

mitch-roblox replied:
Also, why a specific libarchive version? It seems like there are a lot of custom package installs in this recipe that are going to be very brittle.

mitch-roblox replied:
The version of libarchive I get on a vanilla CentOS 7.2 box is the same version that you're downloading: `3.1.2-10.el7_2`. It seems like you could just add it to the list of packages and get rid of the custom download and install.

Brantone commented on recipes/_deps_centos.rb:
```diff
@@ -0,0 +1,112 @@
+
+cache_dir = Chef::Config[:file_cache_path]
+
+%w(
```
How many of these are defined dependents of another?

hyeomans replied:
This comes from mcrouter build script.

mitch-roblox replied:
There's probably a lot of duplication in terms of dependencies, but this is probably the most robust way to handle it for the moment: just list each package out and let the package manager figure it out. Otherwise you're at the whim of distro maintainers changing dependencies on and breaking your cookbook.

Just my 2 cents.

Brantone commented on recipes/_deps_centos.rb:
```diff
@@ -0,0 +1,112 @@
+  not_if "rpm -qa | grep -qx 'libarchive-3.1.2-10.el7_2.x86_6'"
+end
+
+remote_file "#{cache_dir}/automake-1.15-4.fc23.noarch.rpm" do
```
Confusing since this is in the list above on line 61

hyeomans replied:
I wish I had an answer for this, if I remove `cmake` from the package list, `wangle` fails to compile.
`mcrouter` lists cmake as a dependecy in its build scripts, but then I found, through a fork, that adding this specific version, wangle could compile.

Here [is the fork](https://github.com/gdubicki/mcrouter/blob/master/mcrouter/scripts/install_centos_7.2.sh#L19)

Here is the [build script](https://github.com/facebook/mcrouter/blob/master/mcrouter/scripts/install_centos_7.2.sh#L21)

Here is how I discovered [this issue](https://github.com/facebook/mcrouter/issues/187)

mitch-roblox replied:
cmake isn't in the above package list, though. It seems odd to install it explicitly using `yum_package` below. Have you tried adding it to the above package list?

hyeomans replied:
Yeah, it didn't work. I will try again but it will take around 15 minutes to fail.

mitch-roblox replied:
What failed specifically?  The versions of automake and cmake you are installing manually are the same ones I get from yum. They should be the exact same packages.

Brantone commented on recipes/initial_setup.rb:
```diff
@@ -16,24 +17,13 @@
+  shell '/bin/false'
+end
 
-    it 'enables and starts the mcrouter Upstart service' do
```
With this file moved, that means this test is lost. Reason?

hyeomans commented:
* I added `checksum` to `remote_file` resources.
* Added required attributes for folly and wangle, and updated CHANGELOG.

@Brantone @mitch-roblox

mitch-roblox commented on recipes/_deps_centos.rb:
```diff
@@ -0,0 +1,112 @@
+  not_if "rpm -qa | grep -qx 'cmake-2.8.12.2-2.el7.x86_64'"
+end
+
+magic_shell_environment 'CC' do
```
Are the devtoolset-4 versions of gcc/g++ actually needed? And if so, why is regular gcc/g++ also being installed?

hyeomans replied:
https://github.com/facebook/mcrouter/blob/master/mcrouter/scripts/install_centos_7.2.sh#L7-L14

hyeomans replied:
If I remove them, it fails to compile

mitch-roblox replied:
Have you tried remove just plain gcc/g++? Just guessing here, but the devtools-4 version is needed because something won't compile with GCC version 4 (the default for CentOS) and you need version 5. But since you're overriding `CC` and `CXX` to use the devtools-4 install, it seems the plain gcc/g++ are unnecessary.

mitch-roblox commented on README.md:
```diff
@@ -1,17 +1,32 @@
+
+The following platforms are supported and tested with Test Kitchen:
+
+* Ubuntu 14.04
```
This is out-of-date.

hyeomans replied:
Will update

mitch-roblox commented on metadata.rb:
```diff
@@ -4,15 +4,20 @@
-depends 'apt'
-depends 'build-essential'
-depends 'ark', '~> 3.1'
+depends 'apt', '~> 6.1'
```
Small nit, but can you alphabetize the depends? It makes it easier to visually scan.

hyeomans replied:
Updated

mitch-roblox commented on recipes/_deps_ubuntu_16.04.rb:
```diff
@@ -0,0 +1,46 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+include_recipe 'apt'
```
I don't think you need the `apt` cookbook or recipe any more. The `apt_update` resource is included in Chef by default in versions >= 12.7.

All you need to do is add a call to `apt_update 'update'` instead of invoking the `apt` recipe.

mitch-roblox commented on recipes/double_conversion.rb:
```diff
@@ -0,0 +1,49 @@
+  notifies :run, 'execute[cmake]'
+end
+
+execute 'cmake' do
```
This comment applies to a lot of the `execute` blocks, but I would recommend giving them descriptive names (like `compile_double_conversion` and `build_and_install_double_conversion`), not using the name of the command to execute. The reason why is that you have multiple resources in this cookbook that have the same name (like 'cmake') which really makes the resource collection confusing and probably prone to bugs.

hyeomans commented:
I removed all unnecessary dependencies to the minimum required. The only one that is required outside of package resource is `automake-1.15-4.fc23.noarch.rpm` for CentOS, which is also part of the build scripts. @mitch-roblox

Brantone commented on recipes/configure.rb:
```diff
@@ -3,6 +3,7 @@
 # Recipe:: configure
 #
 # Copyright 2015 EverTrue, Inc.
+# Copyright 2017 Roblox, Inc.
```
Do we need to verify this with Legal?

Brantone commented on recipes/default.rb:
```diff
@@ -16,9 +17,13 @@
+include_recipe 'mcrouter::_deps_centos' if node['platform'] == 'centos'
+include_recipe 'mcrouter::_deps_ubuntu_16.04' if node['platform'] == 'ubuntu' && node['platform_version'].to_f	== 16.04
 include_recipe 'memcached::default' if node['mcrouter']['local_memcached']
+include_recipe 'mcrouter::double_conversion'
```
Looking at `double_conversion.rb` recipe, it calls cmake; however, that's in the dependent recipes ... so what will happen if someone just adds `mcrouter::double_conversion` to a run-list? Will it fail because cmake isn't installed?

hyeomans replied:
Yeah, I think it will fail. What's the convention when this happens?

Brantone replied:
Depends on what the intention is ... do we even _want_ to allow it to be in the run-list? If not, can update it to have a leading underscore to imply it shouldn't be in a run-list.

hyeomans replied:
Got it, will update.

Brantone commented on recipes/double_conversion.rb:
```diff
@@ -0,0 +1,56 @@
+end
+
+execute 'Double-Conversion - Build and install' do
+  command 'make && make install'
```
Is this is dependent on `cmake` being in $PATH?

hyeomans commented:
@Brantone renamed to _double_conversion.rb and removed Copyright sentence, I'll ask legal before doing that.

Brantone commented on recipes/default.rb:
```diff
@@ -16,9 +16,13 @@
+include_recipe 'mcrouter::_deps_ubuntu_16.04' if node['platform'] == 'ubuntu' && node['platform_version'].to_f	== 16.04
 include_recipe 'memcached::default' if node['mcrouter']['local_memcached']
+include_recipe 'mcrouter::_double_conversion'
+include_recipe 'mcrouter::folly'
```
Same thing for `folly` and `wangle` as for `double_conversion` : are they intended to be allowed to added in a run-list, if not should have leading underscore, otherwise might need recipes updated to ensure things like `autoreconf` exist on $PATH.

Although, that being said, looks like some of those potential problems previously existed ...

hyeomans replied:
yeah it looks like it was like that before but I updated it.

cthayer commented:
@mitch-roblox @Brantone Let's go ahead and push this through.

Hector will follow up in subsequent PR's with improvements including (but not limited to):

- Supporting SysV and SystemD services

daniel-lucio commented:
I have published the RPM for CentOS 7
https://inside-out.xyz/technology/memcached-protocol-router-rpm-for-centos-7.html

RPM for everybody, enjoy if you need it.
